github/git_commits.go

```diff
@@ -88,29 +88,33 @@ type createCommit struct {
 }
 
 // CreateCommit creates a new commit in a repository.
+// commit must not be nil.
 //
 // The commit.Committer is optional and will be filled with the commit.Author
 // data if omitted. If the commit.Author is omitted, it will be filled in with
 // the authenticated user’s information and the current date.
 //
 // GitHub API docs: https://developer.github.com/v3/git/commits/#create-a-commit
 func (s *GitService) CreateCommit(ctx context.Context, owner string, repo string, commit *Commit) (*Commit, *Response, error) {
+	if commit == nil {
+		return nil, nil, fmt.Errorf("commit must be provided")
```

Comment:
I completely agree that library functions should avoid panic during normal processing.

But that same Effective Go section also says "One possible counterexample is during initialization", and I think another counterexample could be for invalid API usage. For example, consider [`(crypto/rand).Int`](https://godoc.org/crypto/rand#Int) function:

```Go
// Int returns a uniform random value in [0, max). It panics if max <= 0.
func Int(rand io.Reader, max *big.Int) (n *big.Int, err error) {
```

Panicking (rather than returning error) on incorrect API usage seems totally appropriate, since the goal is to let the user know that they are using the API incorrectly as soon as possible, so the most drastic option (such as panic) is desired.

Most other places in the standard library that panic on invalid API usage don't have an error return value (e.g., [`(math/big).Int.Sqrt`](https://godoc.org/math/big#Int.Sqrt), so returning a `errors.New("x must be non-negative")` is not even an option there.

However, that is not very applicable in this situation.

The biggest reason why I think we have to go with returning error here in go-github (rather than panicking) is the previous behavior of this method. Previous behavior was to accept `nil` value for `*Commit` and not panic (but return an error that comes from GitHub response), so it would be completely unacceptable for us to suddenly change that.

If we were designing a new API from scratch, then panicking on incorrect API usage, or choosing to use a value to make it impossible to misuse the API, is something we could've considered. But we're not.

So 👍 to returning error. :)